### PR TITLE
group_info C++ refactor

### PR DESCRIPTION
**Pull request by bhroam**
#### Describe Bug or Feature
Refactor the fairshare structures into C++

#### Describe Your Change
Refactor fairshare_head and group_info into C++.  Eliminate group_path in favor of a vector.

#### Attach Test and Valgrind Logs/Output
Description: Tests from given sources on platforms UBUNTU1804, SUSE15, UBUNTU1604, CENTOS8, RHEL8, CENTOS7, RHEL7, SLES12, SLES15, WIN2K16
Platforms: UBUNTU1804,SUSE15,UBUNTU1604,CENTOS8,RHEL8,CENTOS7,RHEL7,SLES12,SLES15,WIN2K16
Profile: regression
|ID|TOTAL|FAIL|ERROR|TIMEDOUT|SKIP|PASS|
|--|--|--|--|--|--|--|
|7689|4178|0|0|0|15|4163|

### CODE REVIEW

**Review by arungrover — APPROVED** [review]
LGTM

**Review by vstumpf — CHANGES_REQUESTED** [review]

**Comment by bhroam** [comment]
@vstumpf @arungrover I've addressed comments.  Please take another look.

File: src/scheduler/fairshare.cpp [hunk 1]
```diff
@@ -919,13 +917,23 @@ fairshare_head::fairshare_head()
...
 fairshare_head::fairshare_head(fairshare_head& ofhead)
 {
 	last_decay = ofhead.last_decay;
 	root = dup_fairshare_tree(ofhead.root, NULL);
+}
 
+/**
+ * @brief copy assignment operator for fairshare_head
+ */
+fairshare_head fairshare_head::operator=(fairshare_head& ofhead)
```
Comment (arungrover) [inline src/scheduler/fairshare.cpp:931]: Shouldn't this return a reference rather than the object?

**Review by arungrover — COMMENTED** [review]
Just one comment

File: src/scheduler/fairshare.cpp (hunk 1 above)
Comment (bhroam) [reply src/scheduler/fairshare.cpp:931]: Indeed it should.  Thanks for catching this.

**Review by arungrover — APPROVED** [review]
LGTM!

**Comment by bhroam** [comment]
@vstumpf gentle reminder

**Review by vstumpf — APPROVED** [review]
LGTM, sorry it took so long!

**Pull request merged**